### PR TITLE
chore: mint task enhancement

### PR DESCRIPTION
**Pull request by negar-abbasi**
Fixes [#548](https://github.com/algorandfoundation/algokit-cli/issues/548)

## Proposed Changes

  - Ask for mnemonic right after provided account address.
  - Checks values of `total` and `decimals` as per ARC3 if set to True right after it is provided.
  - Retrieve the value of `decimals` and `asset-name` in metadata if given, otherwise prompt the user. 
  - Fixed a mismatch between the decimal value specified in the metadata and the decimal value provided during minting.
  - Option for -n is now only used for networks.
  - Adding a `run_callback_once` for the click option callbacks with prompting.

### CODE REVIEW

File: src/algokit/core/tasks/mint/models.py [hunk 1]
```diff
@@ -74,13 +74,20 @@ def to_file_path(self) -> Path:
             raise ValueError(f"Failed to decode JSON from file {file_path}: {err}") from err
 
     @classmethod
-    def from_json_file(cls: type["TokenMetadata"], file_path: Path | None) -> "TokenMetadata":
+    def from_json_file(cls, file_path: Path | None, name: str | None, decimals: int | None) -> "TokenMetadata":
```
Comment (neilcampbell) [inline src/algokit/core/tasks/mint/models.py:77]: Based on this, it appears that it's possible for name to be None, however the type suggest that it isn't?

File: src/algokit/core/tasks/mint/mint.py [hunk 2]
```diff
@@ -205,11 +197,11 @@ def mint_token(  # noqa: PLR0913
         sp=client.suggested_params(),
         reserve=_reserve_address_from_cid(metadata_cid) if mutable else "",
         unit_name=unit_name,
-        asset_name=asset_name,
+        asset_name=token_metadata.name,
         url=_create_url_from_cid(metadata_cid) + "#arc3" if mutable else "ipfs://" + metadata_cid + "#arc3",
         manager=creator_account.address if mutable else "",
         total=total,
-        decimals=decimals,
+        decimals=token_metadata.decimals,
```
Comment (neilcampbell) [inline src/algokit/core/tasks/mint/mint.py:204]: Do we want to ensure a decimals value is always supplied?

File: src/algokit/cli/tasks/mint.py [hunk 3]
```diff
@@ -86,13 +87,81 @@ def _validate_asset_name(context: click.Context, param: click.Parameter, value:
...
+        value (int): The value of the parameter.
+
+    Returns:
+        int: The value of the parameter if it passes the validation.
+    """
+    token_metadata_path = context.params.get("token_metadata_path")
+    if token_metadata_path is not None:
+        token_metadata = TokenMetadata.from_json_file(token_metadata_path, None, None)
+        if token_metadata.decimals != value:
+            raise click.BadParameter("Token metadata JSON and CLI arguments providing decimals amount must be equal!")
```
Comment (neilcampbell) [inline src/algokit/cli/tasks/mint.py:161]:
```suggestion
            raise click.BadParameter("The value for decimals in the metadata JSON must match the decimals argument.")
```

File: src/algokit/cli/tasks/mint.py [hunk 4]
```diff
@@ -86,13 +87,81 @@ def _validate_asset_name(context: click.Context, param: click.Parameter, value:
     """
 
     if len(value.encode("utf-8")) <= MAX_ASSET_NAME_BYTE_LENGTH:
+        token_metadata_path = context.params.get("token_metadata_path")
+        if token_metadata_path is not None:
+            token_metadata = TokenMetadata.from_json_file(token_metadata_path, None, None)
+            if token_metadata.name != value:
```
Comment (neilcampbell) [inline src/algokit/cli/tasks/mint.py:93]: If name is not set in the metadata file, won't this always raise a badparameter error?

File: src/algokit/core/tasks/mint/mint.py (hunk 2 above)
Comment (negar-abbasi) [reply src/algokit/core/tasks/mint/mint.py:204]: There is always a default of 0 for it.

File: src/algokit/core/tasks/mint/models.py [hunk 5]
```diff
@@ -74,13 +74,15 @@ def to_file_path(self) -> Path:
             raise ValueError(f"Failed to decode JSON from file {file_path}: {err}") from err
 
     @classmethod
-    def from_json_file(cls: type["TokenMetadata"], file_path: Path | None) -> "TokenMetadata":
+    def from_json_file(cls, file_path: Path | None, name: str, decimals: int | None = 0) -> "TokenMetadata":
```
Comment (neilcampbell) [inline src/algokit/core/tasks/mint/models.py:77]: Is it ever possible now for decimals to be None? We can potentially switch this to `decimals: int`?

File: src/algokit/cli/tasks/mint.py [hunk 6]
```diff
@@ -72,18 +75,37 @@ def _validate_unit_name(context: click.Context, param: click.Parameter, value: s
         )
 
 
-def _validate_asset_name(context: click.Context, param: click.Parameter, value: str) -> str:
+def _get_and_validate_asset_name(context: click.Context, param: click.Parameter, value: str | None) -> str:
     """
     Validate the asset name by checking if its byte length is less than or equal to a predefined maximum value.
+    If the asset name is not provided, it prompts the user for the name or retrieves it from a metadata file.
```
Comment (neilcampbell) [inline src/algokit/cli/tasks/mint.py:81]:
```suggestion
    If asset name has not been supplied in the metadata file or via an argument a prompt is displayed.
```

File: src/algokit/cli/tasks/mint.py [hunk 7]
```diff
@@ -93,6 +115,75 @@ def _validate_asset_name(context: click.Context, param: click.Parameter, value:
...
+    try:
+        return get_account_with_private_key(value)
+    except Exception as ex:
+        raise click.BadParameter(str(ex)) from ex
+
+
+def _get_and_validate_decimals(context: click.Context, _: click.Parameter, value: int | None) -> int:
+    """
+    Validate the number of decimal places for the token.
+    If decimals is not provieded, it prompts the user for the decimals or retrieves it from a metadata file.
```
Comment (neilcampbell) [inline src/algokit/cli/tasks/mint.py:138]:
```suggestion
    If decimals has not been supplied in the metadata file or via an argument a prompt is displayed.
```

File: src/algokit/cli/tasks/mint.py [hunk 8]
```diff
@@ -103,26 +194,28 @@ def _validate_asset_name(context: click.Context, param: click.Parameter, value:
     prompt="Provide the address or alias of the asset creator",
     help="Address or alias of the asset creator.",
     type=click.STRING,
+    callback=run_callback_once(callback=_get_creator_account),
```
Comment (neilcampbell) [inline src/algokit/cli/tasks/mint.py:197]:
```suggestion
    callback=run_callback_once(_get_creator_account),
```
Reads nicer and is more consistent.

**Review by neilcampbell — APPROVED** [review]

**Pull request merged**